[Caden] iP

cadencjk opened the pull request:
## Duke Assistant
> “By failing to prepare, you are preparing to fail.” – Benjamin Franklin ([Source](https://blog.rescuetime.com/time-management-quotes/))
<p>
Duke Assistant frees your mind of having to remember things you need to do. It's,

- text-based
- intuitive
- ~~FAST~~ *SUPER FAST* to use

All you need to do is,

1. download it from [here](https://github.com/cadencjk/ip).
2. double-click it.
3. add your tasks.
4. let it manage your tasks for you 😉

And it is **FREE!**

Features:

- [X] Managing tasks
- [ ] Managing deadlines (coming soon)
- [ ] Reminders (coming soon)

____________________________________________________________________
If you Java programmer, you can use it to practice Java too. Here's the `main` method
```
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}
```

Commit:
Add Gradle support

cadencjk committed:
Implemented Greet, Echo, Exit function

cadencjk committed:
Implemented Add, List Functions

cadencjk committed:
Implemented Mark as Done function

cadencjk committed:
Implemented ToDos, Events, Deadlines functions

cadencjk committed:
Conducted Automated Text UI Testing

cadencjk committed:
Handling Checked Exceptions

cadencjk committed:
Implemented Delete function

cadencjk committed:
Tidy code with better style

cadencjk committed:
Implement save function

cadencjk committed:
Replace date as Date object

cadencjk committed:
Merge pull request #1 from cadencjk/branch-Level-7

Implement save function

cadencjk committed:
Merge pull request #2 from cadencjk/branch-Level-8

Replace date as Date object

cadencjk committed:
Merge pull request #3 from nus-cs2103-AY2223S1/add-gradle-support

Add Gradle support

cadencjk committed:
Merge branch 'master' of https://github.com/cadencjk/ip

cadencjk committed:
Make the code more OOP

cadencjk committed:
Package classes into duke

cadencjk committed:
Add JUnit tests

cadencjk committed:
Add comments to classes and functions

cadencjk committed:
Comply with coding standards

cadencjk committed:
Implement find keyword feature

cadencjk committed:
Merge branch 'branch-A-JavaDoc'

cadencjk committed:
Merge branch 'branch-A-CodingStandard'

cadencjk committed:
Merge branch 'branch-Level-9'

# Conflicts:
#	src/main/java/duke/Duke.java
#	src/main/java/duke/TaskList.java

cadencjk committed:
Tweak coding style using CheckStyle

danielk0k commented on src/main/java/duke/Ui.java:
```diff
@@ -0,0 +1,27 @@
+package duke;
+
+public class Ui {
+    private String horizontalLine = "____________________________________________________________\n";
```
`horizontalLine` is a constant here hence by coding standards should be in all caps.
```suggestion
    private final String HORIZONTAL_LINE = "____________________________________________________________\n";
```
Similar comment for the `logo` string

danielk0k commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,41 @@
+    public static CommandType parse(String userText) {
+        String commandWord = userText.split(" ", 2)[0];
+
+        if (commandWord.equals("bye")) {
+            return CommandType.BYE;
+        } else if (commandWord.equals("list")) {
+            return CommandType.LIST;
+        } else if (commandWord.equals("mark")) {
+            return CommandType.MARK;
+        } else if (commandWord.equals("unmark")) {
+            return CommandType.UNMARK;
+        } else if (commandWord.equals("todo")) {
+            return CommandType.TODO;
+        } else if (commandWord.equals("deadline")) {
+            return CommandType.DEADLINE;
+        } else if (commandWord.equals("event")) {
+            return CommandType.EVENT;
+        } else if (commandWord.equals("delete")) {
+            return CommandType.DELETE;
+        } else if (commandWord.equals("find")) {
+            return CommandType.FIND;
+        } else {
+            return CommandType.UNABLE;
+        }
```
Little lengthy with so many conditional statements😅 
Maybe you could consider using a hashmap to map the string command to the enum type commands?
This would also help in the future when you may want to add more commands.

danielk0k commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,121 @@
+            String userText = userInput.nextLine();
+            CommandType type = Parser.parse(userText);
+
+            switch (type) {
+            case BYE:
+                ui.reply("Bye. Hope to see you again soon!");
+                isTerminated = true;
+                break;
+            case LIST:
+                ui.reply(tasks.getList());
+                break;
+            case MARK:
```
Could the lines 40-97 be abstracted to its individual commands/ actions?

danielk0k commented on src/test/java/duke/ToDoTest.java:
```diff
@@ -0,0 +1,33 @@
+        }
+    }
+
+    public ToDo makeToDo(String input) throws DukeException {
+        ToDo dummy = new ToDo();
+        dummy.addName(input);
+        return dummy;
+    }
+}
```
Nice way of simulating a user input from the CLI! 👍🏻

danielk0k reviewed (commented):
Great work for week 3! Definitely there is more room for improvement in terms of adhering to coding standard and OOP. Keep it up 😄

cadencjk committed:
Add GUI for Duke

cadencjk committed:
Improve code's OOP

cadencjk committed:
Fix date format bug

cadencjk committed:
Reorganise files into directories

cadencjk committed:
Improve code's OOP

adeearyaa commented on src/main/java/duke/command/AddCommand.java:
```diff
@@ -0,0 +1,59 @@
+     * @return a String which replies to the user
+     */
+    @Override
+    public String runCommand(Ui ui, Storage storage, TaskList taskList) {
```
nice descriptive method names

adeearyaa commented on src/main/java/duke/command/ByeCommand.java:
```diff
@@ -0,0 +1,35 @@
+     * @return a String which replies to the user
+     */
+    @Override
+    public String runCommand(Ui ui, Storage storage, TaskList taskList) {
```
sorry a bit confused as to why the three params in this runCommand function are not used, are they not needed?

adeearyaa commented on src/main/java/duke/command/CommandType.java:
```diff
@@ -0,0 +1,17 @@
+/**
+ * Stores the different types of command words.
+ */
+public enum CommandType {
```
good use of enums

adeearyaa reviewed (commented):
Good readable code, with nice descriptive javadocs as well!

cadencjk committed:
Use assertions to check important assumptions

There are various section of the codes where it is assumed that the
code reply is correct. In some corner cases, the replies given by
the code made my incorrect.

Using assertions define assumptions about the program state so that
the runtime can verify them. May potentially catch bugs.

Let's,
* Ensure that the replies for the following commands: add, delete,
find, mark, unmark are accurate
*  Ensure the deadline and event dates are properly captured

cadencjk committed:
Improve code quality

There were sections of the code that were not up to par with good
code quality. Production code needs to be of high code quality.

Good code quality improves readability, maintenance and makes it
easy to understand.

Let's
* Remove all forms of dead code
* Change all magic literals to named constants
* Ensure style of code is checked by CheckStyle

cadencjk committed:
Merge pull request #4 from cadencjk/branch-A-Assertions

Use assertions to check important assumptions

cadencjk committed:
Merge branch 'master' of https://github.com/cadencjk/ip into branch-A-CodeQuality

# Conflicts:
#	src/main/java/duke/command/DeleteCommand.java
#	src/main/java/duke/command/FindCommand.java
#	src/main/java/duke/command/MarkCommand.java
#	src/main/java/duke/command/UnmarkCommand.java

cadencjk committed:
Merge branch 'branch-A-CodeQuality'

cadencjk committed:
Improve keyword search capabilities

cadencjk committed:
Merge pull request #5 from cadencjk/branch-BCD-Extension

Improve keyword search capabilities

cadencjk committed:
Improve Duke's GUI

cadencjk committed:
Update README.md

cadencjk committed:
Fix JavaFX runtime warning message
